source/agora/consensus/protocol/Nominator.d

```diff
@@ -854,6 +811,11 @@ extern(D):
         in Hash block_hash) @safe nothrow
     {
         const PublicKey K = block_sig.public_key;
+        if (block_sig.height in this.slot_sigs && K in this.slot_sigs[block_sig.height])
+        {
+            log.trace("Signature already collected for public_key {} at height {}", block_sig.public_key, block_sig.height);
+            return false;
+        }
```

Comment:
```D
if (auto sigs = block_sig.height in this.slot_sigs)
{
  if (K in *sigs)
  {
    /* ... */
  }
}
```

Regarding the message, perhaps use "for ourselves" ?

Reply:
Also this should really use the UTXO instead of the public key

Reply:
Actually:
```D
auto sigs = block_sig.height in this.slot_sigs:
if (sigs is null || K in *sigs) {}
```

And remove the redundant lookups below.

Reply:
Updated to use:
```D
if (sigs is null || K in *sigs)
{
    log.trace("Signature already collected for this node at height {}", block_sig.height);
    return false;
}
```

Reply:
I will raise an issue to tackle to use `UTXO` rather than `PublicKey`  in `slot_sigs`.

Reply:
In fact we only need to check if this received signature is stored in the `slot_sigs` or not. We don't need the other signatures at this height. Will revert to use:-
```D
if (block_sig.height in this.slot_sigs && K in this.slot_sigs[block_sig.height])
```